src/collections/vector.h

```diff
@@ -54,19 +90,31 @@ inline bool Vec<T>::is_empty() {
 
 template <typename T>
 void Vec<T>::push(T data) {
-  T* new_buf = new T[this->len + 1];
-  std::copy(this->buf, this->buf + this->len, new_buf);
-  new_buf[this->len] = data;
-
-  delete[] this->buf;
-
-  this->buf = new_buf;
-  this->len++;
+  size_t total = this->len + 1;
+  if (total > this->_capacity) {
+    if (this->_capacity == 0) {
+      this->_capacity = 1;
+    } else {
+      this->_capacity = this->_capacity * 2;
```

Comment:
prefer `<<`

Reply:
```c
int main() {
    int a = 2;
    int a_2 = a * 2;
    int at2 = a + a;
    int am2 = a << 1;
}
```

gcc produces asssembly code
```asm
	movl	$2, -16(%rbp)
	movl	-16(%rbp), %eax
	addl	%eax, %eax
	movl	%eax, -12(%rbp)
	movl	-16(%rbp), %eax
	addl	%eax, %eax
	movl	%eax, -8(%rbp)
	movl	-16(%rbp), %eax
	addl	%eax, %eax
	movl	%eax, -4(%rbp)
	movl	$0, %eax
	popq	%rbp
```
using  `gcc -O0 -S add.c`
shift will also convert to add
divsion also convert to sar in O0.

I think we don't need to do this for performance.

Reply:
`this->_capacity *= 2;`